x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java

```diff
@@ -0,0 +1,402 @@
+        return version.onOrAfter(MINIMUM_COMPATIBLE);
+    }
+
+    public boolean after(MlConfigVersion version) {
+        if (this.isLegacy != version.isLegacy) {
+            return this.isLegacy ? false : true;
+        }
+        return version.id < id;
+    }
+
+    public boolean onOrAfter(MlConfigVersion version) {
+        if (this.isLegacy != version.isLegacy) {
+            return this.isLegacy ? false : true;
+        }
+        return version.id <= id;
+    }
+
+    public boolean before(MlConfigVersion version) {
+        if (this.isLegacy != version.isLegacy) {
+            return this.isLegacy ? true : false;
+        }
+        return version.id > id;
+    }
+
+    public boolean onOrBefore(MlConfigVersion version) {
+        if (this.isLegacy != version.isLegacy) {
+            return this.isLegacy ? true : false;
+        }
+        return version.id >= id;
+    }
+
+    public boolean between(MlConfigVersion lowerInclusive, MlConfigVersion upperExclusive) {
+        if (upperExclusive.onOrBefore(lowerInclusive)) throw new IllegalArgumentException();
+        return onOrAfter(lowerInclusive) && before(upperExclusive);
+    }
+
+    public static MlConfigVersion fromVersion(Version version) {
+        return fromId(version.id);
+    }
+
+    public static Version toVersion(MlConfigVersion mlConfigVersion) {
+        if (mlConfigVersion.id < MINIMUM_COMPATIBLE.id) {
+            throw new IllegalArgumentException("Cannot convert " + mlConfigVersion.id + ". Incompatible version");
+        }
+        return Version.fromId(mlConfigVersion.id);
+    }
+
+    public static MlConfigVersion getMinMlConfigVersion(DiscoveryNodes nodes, String mlConfigVersionNodeAttrName) {
+        return getMinMaxMlConfigVersion(nodes, mlConfigVersionNodeAttrName).v1();
+    }
+
+    public static MlConfigVersion getMaxMlConfigVersion(DiscoveryNodes nodes, String mlConfigVersionNodeAttrName) {
+        return getMinMaxMlConfigVersion(nodes, mlConfigVersionNodeAttrName).v2();
+    }
+
+    public static Tuple<MlConfigVersion, MlConfigVersion> getMinMaxMlConfigVersion(
+        DiscoveryNodes nodes,
+        String mlConfigVersionNodeAttrName
+    ) {
+        MlConfigVersion minMlConfigVersion = MlConfigVersion.CURRENT;
+        MlConfigVersion maxMlConfigVersion = MlConfigVersion.MINIMUM_COMPATIBLE;
+        for (DiscoveryNode node : nodes) {
+            String mlConfigVersionStr = node.getAttributes().get(mlConfigVersionNodeAttrName);
+            MlConfigVersion mlConfigVersion = fromString(mlConfigVersionStr);
+            if (mlConfigVersion.before(minMlConfigVersion)) {
+                minMlConfigVersion = mlConfigVersion;
+            }
+            if (mlConfigVersion.after(maxMlConfigVersion)) {
+                maxMlConfigVersion = mlConfigVersion;
+            }
+        }
+        return new Tuple<>(minMlConfigVersion, maxMlConfigVersion);
+    }
+
+    public static MlConfigVersion fromString(String str) {
```

Comment:
Just to confirm, we don't need to handle the full version format defined by `Version`: https://github.com/edsavage/elasticsearch/blob/1d8b286f153f1339d7f6b89b2c4c5494372e87f7/build-tools/src/main/java/org/elasticsearch/gradle/Version.java#L66-L81 because the legacy versions we've registered above don't include things like `alpha` or `-SNAPSHOT`?

Reply:
As above.. I think it makes sense to use the `Version.fromString` parser here, so yes, it would understand release candidates etc.